core/devmode/src/main/java/io/quarkus/dev/RuntimeUpdatesProcessor.java

```diff
@@ -31,22 +31,33 @@
 
 public class RuntimeUpdatesProcessor implements HotReplacementContext {
     private static final String CLASS_EXTENSION = ".class";
+    private static final Logger log = Logger.getLogger(RuntimeUpdatesProcessor.class.getPackage().getName());
+
     private final DevModeContext context;
     private final ClassLoaderCompiler compiler;
-    private volatile long lastChange = System.currentTimeMillis();
 
     // file path -> isRestartNeeded
     private volatile Map<String, Boolean> watchedFilePaths = Collections.emptyMap();
+
+    /**
+     * A first scan is considered done when we have visited all modules at least once.
+     * This is useful in two ways.
+     * - To make sure that source time stamps have been recorded at least once
+     * - To avoid re-compiling on first run by ignoring all first time changes detected by
+     * {@link RuntimeUpdatesProcessor#checkIfFileModified(Path, Map)} during the first scan.
+     */
+    private volatile boolean firstScanDone = false;
+
+    private final Map<Path, Long> sourceFileTimestamps = new ConcurrentHashMap<>();
     private final Map<Path, Long> watchedFileTimestamps = new ConcurrentHashMap<>();
+    private final Map<Path, Long> classFileChangeTimeStamps = new ConcurrentHashMap<>();
```

Comment:
It's minor but it should be `Timestamps` not `TimeStamps`.

Reply:
Good catch.